### ISSUE TITLE
Decision path through node thresholds

### ISSUE THREAD

**Issue by tlapusan** [tensorflow/decision-forests#111]
Hi,

I would like to ask you about the decision node path in case the node has a categorical or numerical threshold.

What I observed is when I have a categorical node and the threshold condition is met, then the path is going to the left.
<img width="365" alt="cat_1" src="https://user-images.githubusercontent.com/12815158/170650526-37f4d310-d71f-4977-8451-dc05d3ca41da.png">
<img width="350" alt="cat_2" src="https://user-images.githubusercontent.com/12815158/170650532-92aa2f48-0c9e-4ee2-acce-da96208c7717.png">

If the threshold is a numerical one and the condition is met, then the path is going to the right.
<img width="370" alt="num_1" src="https://user-images.githubusercontent.com/12815158/170650723-65e5d566-34cb-4087-880b-f6884db8d6dc.png">
<img width="316" alt="num_2" src="https://user-images.githubusercontent.com/12815158/170650729-dca37317-e1c8-4666-86bf-c0eb1dd0fab6.png">

Is this behavior the intended one ?

**Comment by rstz** [comment]
Hi, this doesn't look correct to me. Do you have a repro for this?

**Comment by tlapusan** [comment]
Hi @rstz,

You can check this colab notebook:
https://colab.research.google.com/drive/1XvsafToHzDQVR5BOOKEVCxsny_P1pRBU?usp=sharing

**Comment by rstz** [comment]
Thank you, I'll have a look

**Comment by rstz** [comment]
Minor update: This looks like a bug in the way integerized values are treated by TF-DF. In your example Sex_label == 1 corresponds to "male" in the dataset, but it corresponds to "female" when inspecting (and drawing) the graph. We are working on a fix.

Small aside: TF-DF supports string categories, so there is no need to convert strings to integers.

**Comment by tlapusan** [comment]
good to know about string categories, I had the data preprocessing step from other model libraries and used it as it is :)

**Comment by Arnold1** [comment]
@rstz @tlapusan is there any disadvantage to use TF-DF string categories compared to converting the string categories into an integer using Target Encoding or One-Hot Encoding?
when is Target Encoding or One-Hot Encoding useful with decision trees if so or does it have any benefits?
what happens if I have two similar websites one has www.cnn.com and cnn.com - would TF-DF string category map to the same?
what happens if number of categories are high?

**Comment by achoum** [comment]
> Is there any disadvantage to use TF-DF string categories compared to converting the string categories into an integer using Target Encoding or One-Hot Encoding?

Generally, using one-hot encoding with decision forests makes the model larger and less accurate model than using [other options](https://github.com/google/yggdrasil-decision-forests/blob/main/documentation/learners.md#categorical_algorithm). For this reason, it is not recommended.

From experience, and depending on the dataset, target encoding is complementary to CART or RANDOM categorical splits.

> what happens if I have two similar websites one has www.cnn.com and cnn.com - would TF-DF string category map to the same?

Both CART and RANDOM splitter learn conditions of the type "attribute in mask". If this is supported by the dataset, they can learn splits such as "site in [www.cnn.com, cnn.com]. On the other hand, one hot encoding can only check on categorical values at a time.

> what happens if the number of categories are high?

There is a risk of overfitting. In this case, RANDOM categorical splits, target propagation, regularization or more advanced techniques are required.

> how does TF-DF supports string categories internally?

It depends on the semantic of the attribute. [Here](https://github.com/tensorflow/decision-forests/blob/main/tensorflow_decision_forests/tensorflow/core.py#L57) is the list of the supported semantics. For a categorical attribute, as mentioned above, three splitter algorithms are available: CART, RANDOM and ONE_HOT. For a categorical set attribute (e.g. a bag of works), [another algorithm](https://arxiv.org/abs/2009.09991) is used.

**Comment by tlapusan** [comment]
> Minor update: This looks like a bug in the way integerized values are treated by TF-DF. In your example Sex_label == 1 corresponds to "male" in the dataset, but it corresponds to "female" when inspecting (and drawing) the graph. We are working on a fix.
> 
> Small aside: TF-DF supports string categories, so there is no need to convert strings to integers.

Hi @rstz, do you know when this issue will be fixed ?

**Comment by tlapusan** [comment]
> Minor update: This looks like a bug in the way integerized values are treated by TF-DF. In your example Sex_label == 1 corresponds to "male" in the dataset, but it corresponds to "female" when inspecting (and drawing) the graph. We are working on a fix.
> 
> Small aside: TF-DF supports string categories, so there is no need to convert strings to integers.

Could you guide me how to find which are the integer values associated by TF-DF for a categorical features ? 
I'm working to integrate the TF-DF in https://github.com/parrt/dtreeviz library for visualisations and I need the integer values.

Thanks.

**Comment by rstz** [comment]
Hi, apologies for not responding, I missed that email. 

We don't have a fix yet, but this is very high on our Todo list.

**Comment by tlapusan** [comment]
sounds good @rstz , thanks

**Comment by rstz** [comment]
I believe this is solved since dtreeviz support has now landed :)

**Comment by Arnold1** [comment]
there is dtreeviz now?

**Comment by rstz** [comment]
https://www.tensorflow.org/decision_forests/tutorials/dtreeviz_colab is a tutorial for using dtreeviz with TF-DF
